File: python/dynamics_norm_svd.py
import numpy as np
from scipy.linalg import eig
import matplotlib.pyplot as plt

# parameters
Omega = 0.05
Gamma = Omega/5

# Build the superoperator
H_s = np.array([[0, Omega], [Omega, 0]]);
L = np.sqrt(Gamma)*np.array([[0, 1], [1, 0]]);
superop = Liouvillian(H_s,[L])

# Finding matrices containing normalised right and left eigenvectors:
D,left,right = eig(superop, left=True)

# initial state in vectorised form
vec_rho_0 = np.array([0,0,0,1], dtype = complex); 

# solution
ts = np.linspace(0, 200, 200); 
vec_rho_t = np.zeros((len(vec_rho_0), len(ts)))

for k in range(len(H_s)**2):
    norm_fac = np.sqrt(left[:,k].T.conjugate().dot(right[:,k]))
    left_k_dag_norm = left[:,k].T.conjugate()/norm_fac
    right_k_norm = right[:,k]/norm_fac
    ak = left_k_dag_norm.dot(vec_rho_0)
    vec_rho_t = vec_rho_t + ak*np.array([right_k_norm*np.exp(D[k]*t) for t in ts]).T

# Populations dynamics
fig, ax = plt.subplots( figsize = (6,2))
ax.plot(ts*Omega, np.real(vec_rho_t[3]), 'k-')
ax.set_xlabel(r'time ($t\Omega$)', usetex = True, fontsize = 10)
ax.set_ylabel(r'$\mathrm{Tr}[\rho(t)\rho_0]$', usetex=True, fontsize = 10);

File: matlab/dynamics_norm_svd.m
hbar = 1;   d=2;   Omega = 0.05;   Gamma = Omega/5;
t = linspace(0, 200, 50); 

% Building the superoperator
H_s = [0 hbar*Omega; hbar*Omega 0];
L = [0 1; 1 0];
I = eye(2);
P = (-1i/hbar)*(kron(I, H_s) - kron(H_s.', I)) ...
    + Gamma*(kron(conj(L), L) - (1/2)*(kron((L'*L).',I) + kron(I, L'*L)));

% Finding matrices containing right and left eigenvectors:
% D is a diagonal matrix of eigenvalues
% Columns of R_mat and L_mat correspond to left and right eigenvectors s.t.
% P*R_mat = R_mat*D and L_mat'*P = D*L_mat' (where ' denotes conjugate transpose)
[R_mat, D, L_mat] = eig(P);

vec_rho_0 = [0 0 0 1]';                              % Initial rho vector
vec_rho_t = zeros(length(vec_rho_0), length(t));    % Initializing solution

for k=1:(d^2)
    % Obtaining normalized eigenvectors and coefficients
    norm_fac = sqrt(L_mat(:,k)'*R_mat(:,k));
    L_k_dag_norm = L_mat(:,k)'/norm_fac;
    R_k_norm = R_mat(:,k)/norm_fac;
    a(k) = L_k_dag_norm*vec_rho_0;                  
    
    % Summing the kth solution 
    vec_rho_t(:,:) = vec_rho_t(:,:) + a(k)*R_k_norm*exp(D(k,k)*t);
end

% Real part plotted, ignoring small imaginary errors of numerical eigensolutions
plot(t, real(vec_rho_t(4,:)), 'ro'); hold on
